pkg/build/controller/factory/factory.go

```diff
@@ -288,6 +332,61 @@ func (lw *buildLW) Watch(resourceVersion string) (watch.Interface, error) {
 	return lw.client.Builds(kapi.NamespaceAll).Watch(labels.Everything(), fields.Everything(), resourceVersion)
 }
 
+// buildDeleteLW is a ListWatcher implementation that watches for builds being deleted
+type buildDeleteLW struct {
+	client ControllerClient
+	store  cache.Store
+}
+
+// List returns an empty list but adds delete events to the store for all Builds that have been deleted but still have pods.
+func (lw *buildDeleteLW) List() (runtime.Object, error) {
+	glog.V(5).Info("Checking for deleted builds")
+	podList, err := lw.client.KubeClient.Pods(kapi.NamespaceAll).List(labels.Everything(), fields.Everything())
+	if err != nil {
+		glog.V(4).Infof("Failed to find any pods due to error %v", err)
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		if len(pod.Labels[buildapi.BuildLabel]) == 0 {
+			continue
+		}
+		glog.V(5).Infof("Found build pod %s/%s", pod.Namespace, pod.Name)
+
+		build, err := lw.client.Client.Builds(pod.Namespace).Get(pod.Labels[buildapi.BuildLabel])
+		if err != nil && !kerrors.IsNotFound(err) {
+			glog.V(4).Infof("Error getting build for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+			return nil, err
+		}
+		if err != nil && kerrors.IsNotFound(err) {
+			build = nil
+		}
+
+		if build == nil {
+			deletedBuild := &buildapi.Build{
+				ObjectMeta: kapi.ObjectMeta{
+					Name:      pod.Labels[buildapi.BuildLabel],
+					Namespace: pod.Namespace,
+				},
+			}
+			glog.V(4).Infof("No build found for build pod %s/%s, deleting pod", pod.Namespace, pod.Name)
```

Comment:
Besides I'm still quite reluctant on the idea of removing builds even when pods don't exists anymore. We have build pruning to handle that, IMHO.